### ISSUE TITLE
dia_matrix should support indexing

### ISSUE THREAD

**Issue by anntzer** [scipy/scipy#6112]
Computing an kth order difference operator as a sparse matrix is simple:
```
In [2]: def dk(n, k): return (sparse.diags([-1, 1], [0, 1], (n, n)) ** k)[:n - k]

In [3]: dk(5, 2).todense()
Out[3]: 
matrix([[ 1., -2.,  1.,  0.,  0.],
        [ 0.,  1., -2.,  1.,  0.],
        [ 0.,  0.,  1., -2.,  1.]])
```
but this fails for k=1 (and k=0 too):
```
In [4]: dk(5, 1).todense()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-4-585330d35f61> in <module>()
----> 1 dk(5, 1).todense()

<ipython-input-2-655b0e1f2caf> in dk(n, k)
----> 1 def dk(n, k): return (sparse.diags([-1, 1], [0, 1], (n, n)) ** k)[:n - k]

TypeError: 'dia_matrix' object is not subscriptable
```
Given that taking a power already implcitly converts the matrix to CSR format I think it would be reasonable for the indexing to do the same.

**Comment by perimosocordiae** [comment]
This is a bit of a gray area in API design, but to me it seems like math operations are "expensive", so changing sparse formats is acceptable, but indexing should be "cheap", to the point that a silent format change would be surprising (in a bad way) to users.

Besides, the fix for the general case in your function is easy:

    def dk(n, k):
       return (sparse.diags([-1, 1], [0, 1], (n, n), format='csr') ** k)[:n - k]

**Comment by anntzer** [comment]
Your fix still fails for k=0.  (Yes, I know, I can make an explicit call to `tocsr` after the power.)

**Comment by perimosocordiae** [comment]
True, and the behavior that `x ** 0` always makes a DIA-format result is a little surprising.

Regarding your original question, it might make sense to add indexing support for the DIA format, but it shouldn't route through another format.

**Comment by anntzer** [comment]
But that can't really work right?  Say I slice np.eye(n) by taking every other row, the result would be rather ugly in DIA format.
Please reopen the issue, editing the title if you prefer.

**Comment by perimosocordiae** [comment]
Sure, not all indexing operations make sense for DIA format. Slicing one or more contiguous rows/columns should be fairly easy to do, though.

I'm not sure whether to keep this closed as "works as intended", or re-opening as a feature request. @pv, your thoughts?

**Comment by pv** [comment]
As the indexing returns copies, it's probably valid to return anything.  IIRC in some cases it even returns dense matrices? (But there probably should be some hierarchy `dia_matrix -> coo_matrix -> dense matrix` with no loops in the graph.)

In principle it would be nice if the user did not have to care about what the actual sparse format is.

**Comment by perimosocordiae** [comment]
Ok, reopening as a more generic feature request.

**Comment by hpaulj** [comment]
My experience is that math operations, especially on `csr` formats, is relatively cheap, and that indexing is expensive, especially in comparison to array indexing.  More than once on Stackoverflow I have suggested speed ups by direct access to the sparse attributes, e.g.: http://stackoverflow.com/a/37015740/901925

Part of what slows down indexing of sparse matrices is that the result may be another sparse matrix, incurring all the overhead of creating a new matrix.

I've more seen complaints about the lack of indexing for the `coo` format than for the `dia` one.
